File: web/src/lib/types/disk/disk.ts
/**
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * Copyright (c) 2025 The FreeBSD Foundation.
 *
 * This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
 * of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
 * under sponsorship from the FreeBSD Foundation.
 */

import { z } from 'zod/v4';

export const DeviceInfoSchema = z.object({
	name: z.string(),
	info_name: z.string(),
	type: z.string(),
	protocol: z.string()
});

export const PartitionSchema = z.object({
	uuid: z.string(),
	name: z.string(),
	usage: z.string(),
	size: z.number()
});

export const ATASmartAttributeSchema = z.object({
	page: z.number().default(0),
	id: z.number(),
	name: z.string(),
	value: z.number().optional(),
	worst: z.number().optional(),
	thresh: z.number().optional(),
	raw_value: z.number(),
	raw_string: z.string(),
	state: z.number().default(0),
	when_failed: z.string().default(''),
	pre_failure: z.boolean().default(false),
	online: z.boolean().default(false),
	performance: z.boolean().default(false),
	error_rate: z.boolean().default(false),
	event_count: z.boolean().default(false),
	auto_keep: z.boolean().default(false)
});

export const SCSISelfTestResultSchema = z.object({
	type: z.string(),
	status: z.string(),
	lifetime_hours: z.number(),
	lba: z.number(),
	lba_valid: z.boolean(),
	sense_key: z.number(),
	asc: z.number(),
	ascq: z.number()
});

export const SmartDataSchema = z.object({
	device: DeviceInfoSchema,
	passed: z.boolean(),
	health_known: z.boolean().default(false),
	checksum_valid: z.boolean().default(false),
	power_on_hours: z.number(),
	power_cycle_count: z.number(),
	temperature: z.number(),
	self_test_status: z.object({ status: z.string(), remaining_pct: z.number() }).optional(),
	smart_capability: z.number().default(0),
	scsi_self_test_results: z.array(SCSISelfTestResultSchema).optional(),
	attributes: z.array(ATASmartAttributeSchema).nullable().optional()
});

export const NvmeCriticalWarningStateSchema = z.object({
	availableSpare: z.number(),
	temperature: z.number(),
	deviceReliability: z.number(),
	readOnly: z.number(),
	volatileMemoryBackup: z.number()
});

export const SmartNVMeSchema = z.object({
	device: DeviceInfoSchema,
	passed: z.boolean(),
	health_known: z.boolean().default(false),
	power_on_hours: z.number(),
	power_on_hours_exact: z.string().default(''),
	power_cycle_count: z.number(),
	power_cycle_count_exact: z.string().default(''),
	temperature: z.number(),

	criticalWarning: z.string(),
	criticalWarningState: NvmeCriticalWarningStateSchema,
	availableSpare: z.number(),
	availableSpareThreshold: z.number(),
	percentageUsed: z.number(),
	dataUnitsRead: z.number(),
	dataUnitsReadExact: z.string().default(''),
	dataUnitsWritten: z.number(),
	dataUnitsWrittenExact: z.string().default(''),
	hostReadCommands: z.number(),
	hostReadCommandsExact: z.string().default(''),
	hostWriteCommands: z.number(),
	hostWriteCommandsExact: z.string().default(''),
	controllerBusyTime: z.number(),
	controllerBusyTimeExact: z.string().default(''),
	unsafeShutdowns: z.number(),
	unsafeShutdownsExact: z.string().default(''),
	mediaErrors: z.number(),
	mediaErrorsExact: z.string().default(''),
	errorInfoLogEntries: z.number(),
	errorInfoLogEntriesExact: z.string().default(''),
	warningCompositeTempTime: z.number(),
	errorCompositeTempTime: z.number(),
	temperature1TransitionCnt: z.number(),
	temperature2TransitionCnt: z.number(),
	totalTimeForTemperature1: z.number(),
	totalTimeForTemperature2: z.number()
});

export const DiskSchema = z.object({
	uuid: z.string(),
	identityStable: z.boolean().default(false),
	device: z.string(),
	type: z.string(),
	usage: z.string(),
	size: z.number(),
	model: z.string(),
	serial: z.string(),
	gpt: z.boolean(),
	smartData: z.union([SmartNVMeSchema, SmartDataSchema, z.null()]).optional(),
	wearOut: z.string(),
	partitions: z.array(PartitionSchema).default([])
});

export const DiskActionSchema = z.object({
	device: z.string()
});

export const SmartSelfTestKindSchema = z.enum([
	'offline',
	'default',
	'short',
	'extended',
	'conveyance',
	'selective',
	'short_captive',
	'extended_captive',
	'conveyance_captive',
	'selective_captive'
]);

export const SmartSelfTestStartKindSchema = z.enum(['short', 'extended', 'conveyance']);

export const SmartSelfTestCapabilitiesSchema = z.object({
	protocol: z.string().default(''),
	scope: z.string().default(''),
	namespace_id: z.number().default(0),
	supported: z.boolean().default(false),
	single_operation: z.boolean().default(false),
	execution_support_known: z.boolean().default(false),
	offline: z.boolean().default(false),
	default: z.boolean().default(false),
	short: z.boolean().default(false),
	extended: z.boolean().default(false),
	conveyance: z.boolean().default(false),
	selective: z.boolean().default(false),
	short_captive: z.boolean().optional().default(false),
	extended_captive: z.boolean().optional().default(false),
	conveyance_captive: z.boolean().optional().default(false),
	selective_captive: z.boolean().optional().default(false),
	abort: z.boolean().default(false),
	result_log: z.boolean().default(false),
	progress: z.boolean().default(false),
	offline_duration_minutes: z.number().default(0),
	short_duration_minutes: z.number().default(0),
	extended_duration_minutes: z.number().default(0),
	conveyance_duration_minutes: z.number().default(0)
});

export const SmartSelfTestResultSchema = z.object({
	protocol: z.string().optional().default(''),
	type: z.string().default(''),
	mode: z.string().optional().default(''),
	status: z.string().default(''),
	outcome: z.string().optional().default(''),
	remaining_pct: z.number().default(-1),
	lifetime_hours: z.number().default(0),
	lifetime_hours_exact: z.string().default(''),
	lba: z.number().optional().default(0),
	lba_exact: z.string().default(''),
	lba_valid: z.boolean().optional().default(false),
	nsid: z.number().optional().default(0),
	nsid_valid: z.boolean().optional().default(false),
	segment_num: z.number().optional().default(0),
	sense_key: z.number().optional().default(0),
	asc: z.number().optional().default(0),
	ascq: z.number().optional().default(0),
	status_code_type: z.number().optional().default(0),
	status_code_type_valid: z.boolean().optional().default(false),
	status_code: z.number().optional().default(0),
	status_code_valid: z.boolean().optional().default(false),
	checkpoint: z.number().optional().default(0),
	parameter_code: z.number().optional().default(0),
	vendor_specific: z.number().optional().default(0),
	started_at: z.string().nullable().optional().default(null)
});

export const SmartSelfTestStatusSchema = z.object({
	protocol: z.string().default(''),
	namespace_id: z.number().default(0),
	state: z.string().default('idle'),
	execution_status: z.string().default(''),
	type: z.string().default(''),
	running: z.boolean().default(false),
	progress_pct: z.number().default(-1),
	progress_known: z.boolean().default(false),
	remaining_pct: z.number().default(-1),
	remaining_known: z.boolean().default(false),
	estimated_duration_minutes: z.number().default(0),
	offline_collection_status: z.string().default(''),
	offline_collection_running: z.boolean().default(false),
	checksum_valid: z.boolean().default(true),
	results: z.array(SmartSelfTestResultSchema).default([])
});

export const SmartSelfTestDetailsSchema = z.object({
	device: z.string(),
	capabilities: SmartSelfTestCapabilitiesSchema,
	status: SmartSelfTestStatusSchema
});

export const SmartSelfTestScheduleTypeSchema = z.enum(['short', 'extended']);

export const SmartSelfTestScheduleSchema = z.object({
	id: z.number(),
	diskKey: z.string(),
	device: z.string(),
	model: z.string().default(''),
	serial: z.string().default(''),
	testType: SmartSelfTestScheduleTypeSchema,
	cronExpr: z.string(),
	enabled: z.boolean(),
	queuedAt: z.string().nullable().optional().default(null),
	lastRunAt: z.string().nullable().optional().default(null),
	nextRunAt: z.string().nullable().optional().default(null),
	lastStatus: z.string().default('idle'),
	lastError: z.string().default(''),
	progressPct: z.number().default(-1),
	progressKnown: z.boolean().default(false),
	estimatedMinutes: z.number().default(0)
});

export type SmartAttribute = Record<
	string,
	string | number | boolean | Record<string, string | number | boolean>
>;

export type DeviceInfo = z.infer<typeof DeviceInfoSchema>;
export type ATASmartAttribute = z.infer<typeof ATASmartAttributeSchema>;
export type SmartData = z.infer<typeof SmartDataSchema>;
export type SmartNVMe = z.infer<typeof SmartNVMeSchema>;
export type Disk = z.infer<typeof DiskSchema>;
export type Partition = z.infer<typeof PartitionSchema>;
export type SmartSelfTestKind = z.infer<typeof SmartSelfTestKindSchema>;
export type SmartSelfTestStartKind = z.infer<typeof SmartSelfTestStartKindSchema>;
export type SmartSelfTestCapabilities = z.infer<typeof SmartSelfTestCapabilitiesSchema>;
export type SmartSelfTestResult = z.infer<typeof SmartSelfTestResultSchema>;
export type SmartSelfTestStatus = z.infer<typeof SmartSelfTestStatusSchema>;
export type SmartSelfTestDetails = z.infer<typeof SmartSelfTestDetailsSchema>;
export type SmartSelfTestScheduleType = z.infer<typeof SmartSelfTestScheduleTypeSchema>;
export type SmartSelfTestSchedule = z.infer<typeof SmartSelfTestScheduleSchema>;

export type SmartSelfTestScheduleInput = {
	device: string;
	testType: SmartSelfTestScheduleType;
	cronExpr: string;
	enabled: boolean;
};

File: internal/interfaces/services/disk/disk.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package diskServiceInterfaces

import "context"

type Partition struct {
	UUID  string `json:"uuid"`
	Name  string `json:"name"`
	Usage string `json:"usage"`
	Size  uint64 `json:"size"`
}

type Disk struct {
	UUID        string           `json:"uuid"`
	Device      string           `json:"device"`
	Type        string           `json:"type"`
	Usage       string           `json:"usage"`
	Size        uint64           `json:"size"`
	Model       string           `json:"model"`
	Serial      string           `json:"serial"`
	GPT         bool             `json:"gpt"`
	SmartData   any              `json:"smartData"`
	WearOut     string           `json:"wearOut"`
	SelfTestLog *DiskSelfTestLog `json:"selfTestLog,omitempty"`
	Partitions  []Partition      `json:"partitions"`
}

type DeviceInfo struct {
	Name     string `json:"name"`
	InfoName string `json:"info_name"`
	Type     string `json:"type"`
	Protocol string `json:"protocol"`
}

type DiskSelfTestStatus struct {
	Status       string `json:"status"`
	RemainingPct int    `json:"remaining_pct"`
}

type SmartData struct {
	Device              DeviceInfo              `json:"device"`
	Passed              bool                    `json:"passed"`
	ChecksumValid       bool                    `json:"checksum_valid"`
	PowerOnHours        int                     `json:"power_on_hours"`
	PowerCycleCount     int                     `json:"power_cycle_count"`
	Temperature         int                     `json:"temperature"`
	SelfTestStatus      DiskSelfTestStatus      `json:"self_test_status"`
	SmartCapability     uint64                  `json:"smart_capability"`
	SCSISelfTestResults []DiskSCSISelfTestEntry `json:"scsi_self_test_results,omitempty"`

	Attributes []ATASmartAttribute `json:"attributes"`
}

type DiskSCSISelfTestEntry struct {
	Type          string `json:"type"`
	Status        string `json:"status"`
	LifetimeHours uint64 `json:"lifetime_hours"`
	LBA           uint64 `json:"lba"`
	SenseKey      uint8  `json:"sense_key"`
	ASC           uint8  `json:"asc"`
	ASCQ          uint8  `json:"ascq"`
}

type ATASmartAttribute struct {
	Page   int    `json:"page"`
	ID     int    `json:"id"`
	Name   string `json:"name"`
	Value  int    `json:"value"`
	Worst  int    `json:"worst"`
	Thresh int    `json:"thresh"`

	RawValue  int64  `json:"raw_value"`
	RawString string `json:"raw_string"`

	State       int    `json:"state"`
	WhenFailed  string `json:"when_failed"`
	PreFailure  bool   `json:"pre_failure"`
	Online      bool   `json:"online"`
	Performance bool   `json:"performance"`
	ErrorRate   bool   `json:"error_rate"`
	EventCount  bool   `json:"event_count"`
	AutoKeep    bool   `json:"auto_keep"`
}

type NvmeCriticalWarningState struct {
	AvailableSpare       int `json:"availableSpare"`
	Temperature          int `json:"temperature"`
	DeviceReliability    int `json:"deviceReliability"`
	ReadOnly             int `json:"readOnly"`
	VolatileMemoryBackup int `json:"volatileMemoryBackup"`
}

type SMARTNvme struct {
	Device               DeviceInfo `json:"device"`
	Passed               bool       `json:"passed"`
	PowerOnHours         int        `json:"power_on_hours"`
	PowerOnHoursExact    string     `json:"power_on_hours_exact"`
	PowerCycleCount      int        `json:"power_cycle_count"`
	PowerCycleCountExact string     `json:"power_cycle_count_exact"`
	Temperature          int        `json:"temperature"`

	CriticalWarning           string                   `json:"criticalWarning"`
	CriticalWarningState      NvmeCriticalWarningState `json:"criticalWarningState"`
	AvailableSpare            int                      `json:"availableSpare"`
	AvailableSpareThreshold   int                      `json:"availableSpareThreshold"`
	PercentageUsed            int                      `json:"percentageUsed"`
	DataUnitsRead             int                      `json:"dataUnitsRead"`
	DataUnitsReadExact        string                   `json:"dataUnitsReadExact"`
	DataUnitsWritten          int                      `json:"dataUnitsWritten"`
	DataUnitsWrittenExact     string                   `json:"dataUnitsWrittenExact"`
	HostReadCommands          int                      `json:"hostReadCommands"`
	HostReadCommandsExact     string                   `json:"hostReadCommandsExact"`
	HostWriteCommands         int                      `json:"hostWriteCommands"`
	HostWriteCommandsExact    string                   `json:"hostWriteCommandsExact"`
	ControllerBusyTime        int                      `json:"controllerBusyTime"`
	ControllerBusyTimeExact   string                   `json:"controllerBusyTimeExact"`
	UnsafeShutdowns           int                      `json:"unsafeShutdowns"`
	UnsafeShutdownsExact      string                   `json:"unsafeShutdownsExact"`
	MediaErrors               int                      `json:"mediaErrors"`
	MediaErrorsExact          string                   `json:"mediaErrorsExact"`
	ErrorInfoLogEntries       int                      `json:"errorInfoLogEntries"`
	ErrorInfoLogEntriesExact  string                   `json:"errorInfoLogEntriesExact"`
	WarningCompositeTempTime  int                      `json:"warningCompositeTempTime"`
	ErrorCompositeTempTime    int                      `json:"errorCompositeTempTime"`
	Temperature1TransitionCnt int                      `json:"temperature1TransitionCnt"`
	Temperature2TransitionCnt int                      `json:"temperature2TransitionCnt"`
	TotalTimeForTemperature1  int                      `json:"totalTimeForTemperature1"`
	TotalTimeForTemperature2  int                      `json:"totalTimeForTemperature2"`
}

type DiskServiceInterface interface {
	GetDiskDevices(ctx context.Context) ([]Disk, error)
	GetSmartData(disk DiskInfo) (any, *DiskSelfTestLog, error)
	GetWearOut(disk any) (float64, error)
	GetDiskSize(device string) (uint64, error)
	DestroyPartitionTable(device string) error
	IsDiskGPT(device string) bool
	RunSelfTest(disk DiskInfo, testType string) error
	AbortSelfTest(disk DiskInfo) error
	GetSelfTestLog(disk DiskInfo) (*DiskSelfTestLog, error)
	GetExtendedSelfTestLog(disk DiskInfo) (*DiskSelfTestLog, error)
	GetErrorLog(disk DiskInfo) (*DiskErrorLog, error)
	GetExtendedErrorLog(disk DiskInfo) (*DiskErrorLog, error)
	GetNVMEErrorLog(disk DiskInfo) (*DiskNVMEErrorLog, error)
	GetSCTStatus(disk DiskInfo) (*DiskSCTStatus, error)
	GetSCTTempHistory(disk DiskInfo) (*DiskSCTTempHistory, error)
	GetLogDirectory(disk DiskInfo) ([]uint8, error)
	GetDeviceStatistics(disk DiskInfo) ([]DiskAttribute, error)
	GetSelectiveSelfTestLog(disk DiskInfo) (*DiskSelfTestLog, error)
	SetSCTFeatureControl(disk DiskInfo, featureCode uint16, state uint16, persistent bool) error
	SetSCTErrorRecoveryControl(disk DiskInfo, read bool, timeLimit uint16) error
}

type DiskSCTStatus struct {
	FormatVersion     uint16 `json:"format_version"`
	SCTVersion        uint16 `json:"sct_version"`
	SCTSpec           uint16 `json:"sct_spec"`
	StatusFlags       uint32 `json:"status_flags"`
	DeviceState       uint8  `json:"device_state"`
	ExtStatusCode     uint16 `json:"ext_status_code"`
	ActionCode        uint16 `json:"action_code"`
	FunctionCode      uint16 `json:"function_code"`
	LBACurrent        uint64 `json:"lba_current"`
	CurrentTemp       int8   `json:"current_temp"`
	MinTempCycle      int8   `json:"min_temp_cycle"`
	MaxTempCycle      int8   `json:"max_temp_cycle"`
	LifetimeMinTemp   int8   `json:"lifetime_min_temp"`
	LifetimeMaxTemp   int8   `json:"lifetime_max_temp"`
	MaxOpLimit        int8   `json:"max_op_limit"`
	OverTempCount     uint32 `json:"over_temp_count"`
	UnderTempCount    uint32 `json:"under_temp_count"`
	SmartStatusPassed bool   `json:"smart_status_passed"`
	MinERCTime        uint16 `json:"min_erc_time"`
}

type DiskSCTTempSample struct {
	Temperature int8 `json:"temperature"`
}

type DiskSCTTempHistory struct {
	SamplingPeriod uint16              `json:"sampling_period"`
	Interval       uint16              `json:"interval"`
	MaxOpLimit     int8                `json:"max_op_limit"`
	OverLimit      int8                `json:"over_limit"`
	MinOpLimit     int8                `json:"min_op_limit"`
	UnderLimit     int8                `json:"under_limit"`
	CBSize         uint16              `json:"cb_size"`
	CBIndex        uint16              `json:"cb_index"`
	Samples        []DiskSCTTempSample `json:"samples"`
}

type DiskNVMEErrorEntry struct {
	ErrorCount  uint64 `json:"error_count"`
	SQID        uint16 `json:"sqid"`
	CommandID   uint16 `json:"command_id"`
	StatusField uint16 `json:"status_field"`
	ParamError  uint16 `json:"param_error"`
	LBA         uint64 `json:"lba"`
	NamespaceID uint32 `json:"namespace_id"`
}

type DiskNVMEErrorLog struct {
	Entries []DiskNVMEErrorEntry `json:"entries"`
}

type DiskErrorEntry struct {
	ErrorData     uint16 `json:"error_data"`
	ExtendedData  uint16 `json:"extended_data"`
	LifetimeHours uint32 `json:"lifetime_hours"`
	LBA           uint64 `json:"lba"`
	Status        uint8  `json:"status"`
	Error         uint8  `json:"error"`
	SectorCount   uint8  `json:"sector_count"`
	Device        uint8  `json:"device"`
}

type DiskErrorLog struct {
	Entries       []DiskErrorEntry `json:"entries"`
	ChecksumValid bool             `json:"checksum_valid"`
}

type DiskSelfTestEntry struct {
	Type          string `json:"type"`
	Status        string `json:"status"`
	RemainingPct  int    `json:"remaining_pct"`
	LifetimeHours uint64 `json:"lifetime_hours"`
	LBA           uint64 `json:"lba"`
	NSID          uint32 `json:"nsid"`
}

type DiskSelfTestLog struct {
	Entries       []DiskSelfTestEntry `json:"entries"`
	InProgress    bool                `json:"in_progress"`
	ProgressPct   int                 `json:"progress_pct"`
	ChecksumValid bool                `json:"checksum_valid"`
}

type DiskAttribute struct {
	Page      uint32 `json:"page"`
	ID        uint32 `json:"id"`
	Name      string `json:"name"`
	Value     int    `json:"value"`
	Worst     int    `json:"worst"`
	Threshold int    `json:"threshold"`
	RawValue  uint64 `json:"raw_value"`
	RawString string `json:"raw_string,omitempty"`
}
